File: examples/basic_example.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import zmk_studio_api as zmk


def run(client: zmk.StudioClient) -> None:
    print("Lock state:", client.get_lock_state())
    behavior_ids = client.list_all_behaviors()
    print("Behavior count:", len(behavior_ids))

    if behavior_ids:
        details = client.get_behavior_details_bytes(behavior_ids[0])
        print("First behavior details bytes:", len(details))

    keymap_bytes = client.get_keymap_bytes()
    print("Keymap bytes:", len(keymap_bytes))

    layouts_bytes = client.get_physical_layouts_bytes()
    print("Physical layouts bytes:", len(layouts_bytes))

    # Demonstrate typed behavior get/set at (layer 0, position 0).
    before = client.get_key_at(0, 0)
    print("Before:", before)

    client.set_key_at(0, 0, zmk.KeyPress(zmk.Keycode.A))
    after = client.get_key_at(0, 0)
    print("After:", after)

    # Avoid persisting the example mutation.
    if client.check_unsaved_changes():
        client.discard_changes()
        print("Discarded changes")


def main() -> int:
    parser = argparse.ArgumentParser(description="Basic zmk_studio_api example")
    sub = parser.add_subparsers(dest="transport", required=True)

    serial = sub.add_parser("serial", help="Connect over serial")
    serial.add_argument("port", help="Serial port path (for example COM3)")

    ble = sub.add_parser("ble", help="Connect over BLE")

    args = parser.parse_args()

    if args.transport == "serial":
        client = zmk.StudioClient.open_serial(args.port)
    else:
        devices = zmk.StudioClient.list_ble_devices()
        if not devices:
            raise SystemExit("No Bluetooth keyboards found. Pair/connect your keyboard first.")

        device_id, local_name = devices[0]
        print("Using BLE device:", local_name or device_id)
        client = zmk.StudioClient.open_ble(device_id)

    run(client)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/basic_example.rs
use std::error::Error;
use std::io::{Read, Write};
use std::process::ExitCode;

#[cfg(feature = "ble")]
use zmk_studio_api::transport::ble::BleTransport;
#[cfg(feature = "serial")]
use zmk_studio_api::transport::serial::SerialTransport;
use zmk_studio_api::{Behavior, ClientError, HidUsage, Keycode, StudioClient};

fn main() -> ExitCode {
    match run() {
        Ok(()) => ExitCode::SUCCESS,
        Err(err) => {
            eprintln!("error: {err}");
            ExitCode::from(1)
        }
    }
}

fn run() -> Result<(), Box<dyn Error>> {
    let mut args = std::env::args().skip(1);
    let Some(mode) = args.next() else {
        print_usage();
        return Ok(());
    };

    match mode.as_str() {
        "serial" => {
            #[cfg(feature = "serial")]
            {
                let Some(port) = args.next() else {
                    print_usage();
                    return Ok(());
                };
                let client = StudioClient::new(SerialTransport::open(&port)?);
                run_example(client)
            }
            #[cfg(not(feature = "serial"))]
            {
                Err("built without `serial` feature".into())
            }
        }
        "ble" => {
            #[cfg(feature = "ble")]
            {
                let _name_substring = args.next();
                let client = StudioClient::new(BleTransport::connect_first()?);
                run_example(client)
            }
            #[cfg(not(feature = "ble"))]
            {
                Err("built without `ble` feature".into())
            }
        }
        _ => {
            print_usage();
            Ok(())
        }
    }
}

fn run_example<T: Read + Write>(mut client: StudioClient<T>) -> Result<(), Box<dyn Error>> {
    let info = client.get_device_info()?;
    println!("Device: {}", info.name);
    println!("Lock: {:?}", client.get_lock_state()?);

    let behavior_ids = client.list_all_behaviors()?;
    println!("Behavior count: {}", behavior_ids.len());
    if let Some(first_behavior_id) = behavior_ids.first().copied() {
        let details = client.get_behavior_details(first_behavior_id)?;
        println!("First behavior: {} ({})", details.id, details.display_name);
    }

    let keymap = match client.get_keymap() {
        Ok(keymap) => keymap,
        Err(ClientError::Meta(_)) => {
            println!("Keymap request denied (likely locked); press `&studio_unlock` then rerun.");
            return Ok(());
        }
        Err(err) => return Err(Box::new(err)),
    };
    println!("Layers: {}", keymap.layers.len());

    let layouts = client.get_physical_layouts()?;
    println!(
        "Physical layouts: {} (active index: {})",
        layouts.layouts.len(),
        layouts.active_layout_index
    );

    let Some(first_layer) = keymap.layers.first() else {
        return Ok(());
    };
    if first_layer.bindings.is_empty() {
        return Ok(());
    }

    let layer_id = first_layer.id;
    let key_position = 0;

    let before = client.get_key_at(layer_id, key_position)?;
    println!("Before: {before:?}");

    client.set_key_at(
        layer_id,
        key_position,
        Behavior::KeyPress(HidUsage::from_encoded(Keycode::A.to_hid_usage())),
    )?;
    let after = client.get_key_at(layer_id, key_position)?;
    println!("After:  {after:?}");

    // Change management APIs.
    let has_changes = client.check_unsaved_changes()?;
    println!("Unsaved changes: {has_changes}");
    if has_changes {
        client.discard_changes()?;
    }

    Ok(())
}

fn print_usage() {
    println!("Usage:");
    println!("  cargo run --example basic_example -- serial <PORT>");
    println!("  cargo run --example basic_example --features ble -- ble");
}
